machinelearning/vllm.md

```diff
@@ -0,0 +1,74 @@
+# Large Language Model (LLM) inference on Graviton CPUs with vLLM
+
+**Introduction**
+
+vLLM is a fast and easy-to-use library for LLM inference and serving. It provides an OpenAI-compatible API server and support NVIDIA GPUs, CPUs and AWS Neuron. vLLM has been adapted to work on ARM64 CPUs with NEON support, leveraging the CPU backend initially developed for the x86 platform. ARM CPU backend currently supports Float32, FP16 and BFloat16 datatypes.
+This document covers how to build and run vLLM for LLM inference on AWS Graviton based Amazon EC2 Instances. 
+
+# How to use vLLM on Graviton CPUs
+
+There are no pre-built wheels or images for Graviton CPUs, so you must build vLLM from source.
+
+**Prerequisites**
+
+Graviton3(E) (e.g. *7g instances) and Graviton4 (e.g. *8g instances) CPUs support BFloat16 format and MMLA instructions for machine learning (ML) acceleration. These hardware features are enabled starting with Linux Kernel version 5.10. So, it is highly recommended to use the AMIs based on Linux Kernel 5.10 and beyond for the best LLM inference performance on Graviton Instances. Use the following queries to list the AMIs with the recommended Kernel versions. New Ubuntu 22.04, 24.04, and AL2023 AMIs all have kernels newer than 5.10.
+
+The following steps were tested on a Graviton3 R7g.2xlarge and Ubuntu 24.04.1
+
+```
+# For Kernel 5.10 based AMIs list
+aws ec2 describe-images --owners amazon --filters "Name=architecture,Values=arm64" "Name=name,Values=*kernel-5.10*" --query 'sort_by(Images, &CreationDate)[].Name'
+
+# For Kernel 6.x based AMIs list
+aws ec2 describe-images --owners amazon --filters "Name=architecture,Values=arm64" "Name=name,Values=*kernel-6.*" --query 'sort_by(Images, &CreationDate)[].Name'
+```
+
+**Install Compiler and Python packages**
+```
+sudo apt-get update  -y
+sudo apt-get install -y gcc-12 g++-12 libnuma-dev python3-dev
+sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-12 10 --slave /usr/bin/g++ g++ /usr/bin/g++-12
+```
+
+**Create a new Python environment using uv, a very fast Python environment manager**
+```
+curl -LsSf https://astral.sh/uv/install.sh | sh
+uv venv venv --python 3.12 --seed
+source venv/bin/activate
+```
```

Comment:
The formatting with nested back ticks isn't working, so the suggested change removes the back ticks, but that was not my intention.

```suggestion
**Create a new Python virtual environment**

virtualenv venv
source venv/bin/activate
```